lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""OpenSearch Plugin model.
+
+In OpenSearch, plugins are also called interchangeably as extensions and modules.
+
+A plugin configuration footprint is composed of:
+* Jar files installed in one of the
+  ${OPENSEARCH_HOME}/plugins
+  ${OPENSEARCH_HOME}/modules
+* Configuration passed to the main opensearch.yml, must be removed at plugin removal
+* Secrets stored in the keystore
+* The file: plugin.properties
+* And the security policy in: security.policy
+* Other plugins which it depends upon
+
+One last piece of the configuration is any index data uploaded to the cluster using
+OpenSearch APIs. That last bit of data must be done by inherinting the OpenSearchPlugin
+class and implementing the necessary extra logic.
+
+
+This file implements abstract methods and data that are common to every plugin during
+its lifecycle, including methods to manage configuration files, main processes (install,
+upgrade, uninstall, etc).
+
+The plugin lifecycle runs through the following steps:
+
+INSTALL > is_installed > is_enabled > needs_upgrade > upgrade > UNINSTALL
+
+The meaning of each step is, as follows:
+* is_installed: the installation happened correctly and the JAR files are set
+* is_enabled: all the configurations have been applied and restart is done, if needed
+* needs_upgrade: once the main OpenSearch is upgraded, the plugin needs to check if an
+                 upgrade is also needed or not.
+* upgrade: run the necessary actions to upgrade the plugin
+
+========================================================================================
+
+                             STEPS TO ADD A NEW PLUGIN
+
+========================================================================================
+
+
+For a new plugin, add the plugin to the list of "OpenSearchPluginsAvailable" available
+in opensearch_plugin_manager.py and override the abstract OpenSearchPlugin.
+
+Add a new configuration in the config.yaml with "plugin_" as prefix to its name.
+Add the corresponding config-name to the OpenSearchPluginsAvailable.
+
+If a given plugin depends on a relation, e.g. repository-s3, then add relation name
+as well. For example:
+    OpenSearchPluginsAvailable = {
+        ...
+        "opensearch-knn": {
+            "class": OpenSearchPlugin,
+            "config-name": "plugin_opensearch_knn",
+            "relation-name": ""
```

Comment:
What would be the use of this option?

Reply:
@Mehdi-Bendriss it is the same pattern as: https://github.com/canonical/postgresql-operator/blob/main/config.yaml#L5